File: src/types/vault.ts
export interface VaultMeta {
  id: string;
  name: string;
  path: string;
  lastOpened: number;
  isEncrypted: boolean;
}

export interface NoteFile {
  relativePath: string;
  name: string;
  extension: string;
  modifiedAt: number;
  size: number;
  isFolder: boolean;
  children?: NoteFile[];
}

export interface NoteContent {
  content: string;
  hash: string;
  modifiedAt: number;
}

export interface WriteResult {
  hash: string;
  conflict?: ConflictInfo;
}

export interface ConflictInfo {
  ourContent: string;
  theirContent: string;
  relativePath: string;
}

export type MemberRole = 'viewer' | 'editor' | 'admin';

export interface VaultMember {
  userId: string;
  userName: string;
  role: MemberRole;
}

export interface VaultConfig {
  id: string;
  name: string;
  knownUsers: KnownUser[];
  owner?: string;
  members?: VaultMember[];
  isEncrypted?: boolean;
}

export interface KnownUser {
  userId: string;
  userName: string;
  userColor: string;
  lastSeen: number;
}

File: src-tauri/src/models/vault.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct VaultMeta {
    pub id: String,
    pub name: String,
    pub path: String,
    pub last_opened: u64,
    /// Whether the vault files are encrypted at rest.
    #[serde(default)]
    pub is_encrypted: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum MemberRole {
    Viewer,
    Editor,
    Admin,
}

impl Default for MemberRole {
    fn default() -> Self {
        MemberRole::Editor
    }
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct VaultMember {
    pub user_id: String,
    pub user_name: String,
    pub role: MemberRole,
}

#[derive(Debug, Serialize, Deserialize, Clone, Default)]
#[serde(rename_all = "camelCase")]
pub struct VaultConfig {
    pub id: String,
    pub name: String,
    pub known_users: Vec<KnownUser>,
    /// UserId of the vault owner (the user who created it).
    #[serde(default)]
    pub owner: Option<String>,
    /// Explicit role assignments for collaborators.
    #[serde(default)]
    pub members: Vec<VaultMember>,
    /// Reserved for future encryption support.
    #[serde(default)]
    pub is_encrypted: bool,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct KnownUser {
    pub user_id: String,
    pub user_name: String,
    pub user_color: String,
    pub last_seen: u64,
}
